homeworks/week5/Minh-Le.java

```diff
@@ -0,0 +1,65 @@
+// Find lowest common boss in a tree;
+// Input: [1, 3, 3, -1, -1, 4, 2, 8, 3]
+// 0 reports to 1, 1 reports to 3, 2 reports to 3, 3 reports to -1 (highest boss), etc.
+//      -1
+//     /  \
+//    3    4
+//  / | \   \
+// 8  2  1   5
+// |  |  |
+// 7  6  0
+// Assumption: the input array is non-empty. And  and valid inputs.
+// Requirement: No extra storage. No build-in library.
+import java.util.HashMap;
+
+class Main {
+  public static void main(String[] args) {
+    int[] boss = { 1, 3, 3, -1, -1, 4, 2, 8, 3 };
+    System.out.println(lowestCommonBoss(boss, 1, 1)); // 1
+    System.out.println(lowestCommonBoss(boss, 2, 0)); // 3
+    System.out.println(lowestCommonBoss(boss, 6, 0)); // 3
+    System.out.println(lowestCommonBoss(boss, 7, 1)); // 3
+    System.out.println(lowestCommonBoss(boss, 4, 5)); // 4
+    System.out.println(lowestCommonBoss(boss, 6, 5)); // -1
+    System.out.println(lowestCommonBoss(boss, 4, 2)); // -1
+    System.out.println(lowestCommonBoss(boss, 4, 3)); // -1
+    System.out.println(lowestCommonBoss(boss, 4, -1)); // -1
+  }
+
+  static int lowestCommonBoss(int[] boss, int e1, int e2) {
+    if (e1 == e2) {
+      return e1;
+    }
+    if (e1 == -1 || e2 == -1) {
+      return -1;
+    }
+
+    // check index out of range edge cases
+    if (e1 >= boss.length || e2 >= boss.length || e1 < 0 || e2 < 0) {
+      return -1;
+    }
+    HashMap<Integer, Boolean> pathToE1 = new HashMap<>();
```

Comment:
Time complexity: O(logN) ~ O(H) where H is tree height.

Space complexity is O(H) here, see sample solution with O(1) https://repl.it/@HienTo/lowestCommonBoss-Solution#Main.java